anomalib/models/cs_flow/torch_model.py

```diff
@@ -0,0 +1,531 @@
+            y12 = self.exp(s11) * x12 + t11
+            y22 = self.exp(s21) * x22 + t21
+
+        else:  # names of x and y are swapped!
+            r01, r11, r21 = self.subnet1(x01, x11, x21)
+
+            s01, t01 = r01[:, : self.split_len2], r01[:, self.split_len2 :]
+            s11, t11 = r11[:, : self.split_len2], r11[:, self.split_len2 :]
+            s21, t21 = r21[:, : self.split_len2], r21[:, self.split_len2 :]
+
+            y02 = (x02 - t01) / self.exp(s01)
+            y12 = (x12 - t11) / self.exp(s11)
+            y22 = (x22 - t21) / self.exp(s21)
+
+            r02, r12, r22 = self.subnet2(y02, y12, y22)
+
+            s02, t02 = r02[:, : self.split_len2], r01[:, self.split_len2 :]
+            s12, t12 = r12[:, : self.split_len2], r11[:, self.split_len2 :]
+            s22, t22 = r22[:, : self.split_len2], r21[:, self.split_len2 :]
+
+            y01 = (x01 - t02) / self.exp(s02)
+            y11 = (x11 - t12) / self.exp(s12)
+            y21 = (x21 - t22) / self.exp(s22)
+
+        z_dist0 = torch.cat((y01, y02), 1)
+        z_dist1 = torch.cat((y11, y12), 1)
+        z_dist2 = torch.cat((y21, y22), 1)
+
+        z_dist0 = torch.clamp(z_dist0, -1e6, 1e6)
+        z_dist1 = torch.clamp(z_dist1, -1e6, 1e6)
+        z_dist2 = torch.clamp(z_dist2, -1e6, 1e6)
+
+        jac0 = torch.sum(self.log_e(s01), dim=(1, 2, 3)) + torch.sum(self.log_e(s02), dim=(1, 2, 3))
+        jac1 = torch.sum(self.log_e(s11), dim=(1, 2, 3)) + torch.sum(self.log_e(s12), dim=(1, 2, 3))
+        jac2 = torch.sum(self.log_e(s21), dim=(1, 2, 3)) + torch.sum(self.log_e(s22), dim=(1, 2, 3))
+
+        # Since Jacobians are only used for computing loss and summed in the loss, the idea is to sum them here
+        return [z_dist0, z_dist1, z_dist2], torch.stack([jac0, jac1, jac2], dim=1).sum()
+
+    def output_dims(self, input_dims: List[Tuple[int]]):
+        """Output dimensions of the module."""
+        return input_dims
+
+
+class CsFlowModel(nn.Module):
+    """CS Flow Module.
+
+    Args:
+        input_size (Tuple[int, int]): Input image size.
+        n_coupling_blocks (int): Number of coupling blocks.
+        clamp (float): Clamp value for the coupling blocks.
+        num_channels (int): Number of channels in the input image.
+    """
+
+    def __init__(
+        self,
+        input_size: Tuple[int, int],
+        cross_conv_hidden_channels: int,
+        n_coupling_blocks: int = 4,
+        clamp: int = 3,
+        num_channels: int = 3,
+    ):
+
+        super().__init__()
+        self.input_dims = (num_channels, *input_size)
+        self.n_coupling_blocks = n_coupling_blocks
+        self.kernel_sizes = [3] * (n_coupling_blocks - 1) + [5]
+        self.clamp = clamp
+        self.cross_conv_hidden_channels = cross_conv_hidden_channels
+        self.feature_extractor = MultiScaleFeatureExtractor(n_scales=3, input_size=input_size)
+        self.graph = self._create_graph()
+
+    def _create_graph(self):
```

Comment:
Maybe you could place this stuff in a separate class which inherits from `GraphINN`, and which is instantiated from the constructor of the torch model. This would keep the `CsFlowModel` class a bit cleaner.